Add variable selection priors

NathanielF opened the pull request:
Just a draft for the minute. Working through some ideas. 

<!-- readthedocs-preview causalpy start -->
----
📚 Documentation preview 📚: https://causalpy--568.org.readthedocs.build/en/568/

<!-- readthedocs-preview causalpy end -->

drbenvincent commented on causalpy/experiments/instrumental_variable.py:
```diff
@@ -51,6 +51,12 @@ class InstrumentalVariable(BaseExperiment):
+    :param vs_prior_type : str or None, default=None
+        Type of variable selection prior: 'spike_and_slab', 'horseshoe', or None.
+        If None, uses standard normal priors.
+    :param vs_hyperparams : dict, optional
```
We should add that into `AGENTS.md` if it's not already there.

NathanielF replied:
> This is sphinx format and not numpy

You got me. The doc strings were AI generated. Will fix.

NathanielF commented:
Marking this one as ready for review. There is still some work to be done on the notebook illustrating the functionality. But i think there is enough here that's it worth flagging the architecture choices for discussion. I've made the variable selection priors available as a module. Currently, just integrated with the IV class, but in principle can be dropped into all regression based modules with coefficients. The pattern simply requires an if-else block to be used in e.g. the propensity score model, linear regression model etc.... 

What do you guys think?

juanitorduz commented:
Great @NathanielF ! I think the notebooks need a bit more storyline and explanation 🙏

drbenvincent commented:
Will take a look at this. But it's worth updating from `main` first. There are likely to be some merge conflicts

drbenvincent commented:
Added some comments on the new notebook. Not looked at the code yet.

NathanielF commented on docs/source/notebooks/iv_vs_priors.ipynb:
```diff
@@ -0,0 +1,2219 @@
+{
```
<p>Added a note and clarification</p>

NathanielF commented on docs/source/notebooks/iv_vs_priors.ipynb:
```diff
@@ -0,0 +1,2219 @@
+{
```
<p>The general prediction parameters and especially instruments will be heaped at zero but the treatment effect is real as per the simulation and we want the models to identify the true effect here, which they generally do</p>

NathanielF commented on docs/source/notebooks/iv_vs_priors.ipynb:
```diff
@@ -0,0 +1,2219 @@
+{
```
<p>Done</p>

NathanielF commented on causalpy/variable_selection_priors.py:
```diff
@@ -0,0 +1,594 @@
+
+        # Compute inclusion probabilities
+        probabilities = (gamma > threshold).mean(dim="sample").to_array()
+        gamma_mean = gamma.mean(dim="sample").to_array()
```
this is incorrect

juanitorduz commented on causalpy/pymc_models.py:
```diff
@@ -772,50 +855,35 @@ def sample_predictive_distribution(self, ppc_sampler: str | None = "jax") -> Non
+        y,
+        t,
+        coords,
+        priors,
```
type hints missing

NathanielF replied:
fixed this

juanitorduz reviewed (dismissed):
This is awesome @NathanielF ! Just some minor comments :)

NathanielF commented on causalpy/pymc_models.py:
```diff
@@ -677,7 +678,10 @@ def build_model(  # type: ignore
+        priors,
+        vs_prior_type=None,
+        vs_hyperparams=None,
+        binary_treatment=False,
```
added type hints

NathanielF commented:
On the cholesky decomposition @juanitorduz , i  re-parameterised again in the binary case. We refactored the joint likelihood into a Conditional Likelihood. 

Key Changes: 
- Elimination of Cholesky Decomposition: We no longer draw from a multivariate distribution. Instead, we draw the latent treatment error ($V$) and use it to adjust the mean of the outcome equation. We have a conditional formulation now. 

Introduction of the Correction Term: We added expected_U = rho * (sigma_U / sigma_V) * V to the outcome mean. This term "soaks up" the variation in the outcome that is correlated with the treatment assignment

Logistic Error Parameterization: To improve sampling stability and match the invlogit link function, we parameterized $V$ using Inverse Transform Sampling from a Uniform distribution into a Standard Logistic distribution.

cursor etc was pushing me towards an inverse probit formulation, but i just found it much more brittle than the logit.

drbenvincent reviewed (approved):
* "eplore" -> "explore"
* suppress warning output in notebook?
* "cull" too informal for non-native speakers?

Approving, and will leave you to make any minor edits before merging or to see if anyone else had minor change requests

drbenvincent commented:
It would be great to see if we can get these priors working in the pymc-extras `Prior` class way of doing things. In a future PR of course. Could potentially be a good job for an LLM :)
